webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+	Payload       DiscordWebhookDict `json:"payload"`
+	MessageNumber int
+}
+
+func (m *Message) String() string {
+	return fmt.Sprintf("S#: %d Discard: %t Edit: %t Msg#: %d",
+		m.ServerNumber, m.Discardable, m.Edit, m.MessageNumber)
+}
+
+// Tracks bucket specific rate limits
+type RateLimitState struct {
+	BucketID    string
+	Remaining   int
+	Limit       int
+	ResetTime   time.Time
+	ResetAfter  time.Duration
+	RateLimited bool
+	mu          sync.Mutex
+}
+
+func (r *RateLimitState) String() string {
+	return fmt.Sprintf("ID: %s Remaining: %d Limit: %d ResetTime: %s ResetAfter: %d RateLimited: %t",
+		r.BucketID, r.Remaining, r.Limit, r.ResetTime, r.ResetAfter, r.RateLimited)
+}
+
+// Tracks our local rate limit window
+var localRateLimitState struct {
+	Requests int
+	Window   time.Time
+	mu       sync.Mutex
+}
+
+// Manages a rate limit bucket
+type BucketWorker struct {
+	QueueKey  string
+	rdb       *redis.Client
+	ctx       context.Context
+	rateLimit *RateLimitState
+}
+
+func NewBucketWorker(bucketID string, rdb *redis.Client) *BucketWorker {
+	return &BucketWorker{
+		QueueKey:  bucketQueuePrefix + bucketID,
+		rdb:       rdb,
+		ctx:       context.Background(),
+		rateLimit: &RateLimitState{BucketID: bucketID},
+	}
+}
+
+func GetWorker(bucket string, mu *sync.Mutex) (*BucketWorker, bool) {
+	mu.Lock()
+	defer mu.Unlock()
+	worker, exists := workerLookup[bucket]
+	return worker, exists
+}
+
+func SetWorker(bucket string, worker *BucketWorker, mu *sync.Mutex) {
+	mu.Lock()
+	defer mu.Unlock()
+	workerLookup[bucket] = worker
+}
+
+// Keeps us under `HLL_LOCAL_MAX_SENDS_PER_SEC` across all requests
+func CheckLocalRateLimit(requests int, now time.Time) error {
+	localRateLimitState.mu.Lock()
+	defer localRateLimitState.mu.Unlock()
+
+	// If it's been more than 1 second since our last request; reset the window
+	if now.Sub(localRateLimitState.Window) >= time.Second {
+		localRateLimitState.Requests = requests
+		localRateLimitState.Window = now
+	}
+
+	// Don't exceed our configured max requests per second
+	if localRateLimitState.Requests >= localRateLimit {
+		localRateLimitState.mu.Unlock()
+		return fmt.Errorf("local rate limit exceeded")
+	}
+
+	localRateLimitState.Requests++
+	return nil
+}
+
+func GetWorkerRateLimitSleepTime(rateLimit *RateLimitState) time.Duration {
+	rateLimit.mu.Lock()
+	defer rateLimit.mu.Unlock()
+
+	if rateLimit.RateLimited {
+		return time.Until(rateLimit.ResetTime)
+	}
+
+	return 0
+}
+
+// Makes the HTTP request to Discord and updates rate limit state
+func SendWebhook(rdb *redis.Client, msg *Message, rateLimit *RateLimitState) (string, error) {
+	now := time.Now()
+	if err := CheckLocalRateLimit(0, now); err != nil {
+		return "", err
+	}
+
+	// Extract what we need to make our POST request to Discord
+	payload := map[string]any{}
+	if msg.Payload.Content != nil {
+		payload["content"] = *msg.Payload.Content
+	}
+	if len(msg.Payload.Embeds) > 0 {
+		payload["embeds"] = msg.Payload.Embeds
```

Comment:
this assignment can be done without the if condition. An empty slice (or nil for that matter) would mean the same as not having `embeds` added in the first place, iirc.